File: fortress/core/pii_redactor.py
"""
PII Redaction Engine (SSN, Credit Cards, Emails, Phones).
"""

from __future__ import annotations
import re
from typing import Any, Dict, List, Tuple
from fortress.config import OutboundGuardConfig
from fortress.core.models import RiskLevel, ViolationRecord


class PIIRedactor:
    """
    High-performance PII masking engine with Luhn checksum validation.
    """

    PII_PATTERNS: Dict[str, re.Pattern] = {
        "ssn": re.compile(r'\b\d{3}[- ]?\d{2}[- ]?\d{4}\b'),
        "email": re.compile(r'\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}\b'),
        "phone": re.compile(r'(?:\+?1[-. ]?)?\(?([0-9]{3})\)?[-. ]?([0-9]{3})[-. ]?([0-9]{4})\b'),
        "credit_card": re.compile(r'\b[2-6](?:[ -]*?\d){12,18}\b'),
    }

    VALID_CC_PREFIXES = ('2', '3', '4', '5', '6')

    def __init__(self, config: OutboundGuardConfig):
        self.config = config

    def luhn_check(self, card_num: str) -> bool:
        # Fast character extraction without re.sub allocation
        digits = [ord(c) - 48 for c in card_num if '0' <= c <= '9']
        n = len(digits)
        if n < 13 or n > 19:
            return False
        if str(digits[0]) not in self.VALID_CC_PREFIXES:
            return False
        checksum = 0
        parity = n % 2
        for i, d in enumerate(digits):
            if i % 2 == parity:
                doubled = d * 2
                checksum += doubled - 9 if doubled > 9 else doubled
            else:
                checksum += d
        return checksum % 10 == 0

    def redact(self, data: Any) -> Tuple[Any, List[ViolationRecord]]:
        if not self.config.mask_pii:
            return data, []

        violations: List[ViolationRecord] = []
        sanitized = self._process_data(data, violations)
        return sanitized, violations

    def _process_data(self, data: Any, violations: List[ViolationRecord], depth: int = 0, max_depth: int = 20) -> Any:
        if depth > max_depth:
            return data
        if isinstance(data, str):
            return self._mask_text(data, violations)
        elif isinstance(data, dict):
            return {k: self._process_data(v, violations, depth + 1, max_depth) for k, v in data.items()}
        elif isinstance(data, list):
            return [self._process_data(item, violations, depth + 1, max_depth) for item in data]
        return data

    def _mask_text(self, text: str, violations: List[ViolationRecord]) -> str:
        if not text or len(text) < 5:
            return text

        has_digits = any('0' <= c <= '9' for c in text)
        has_at = '@' in text

        # Fast exit if text contains neither digits nor '@'
        if not has_digits and not has_at:
            return text

        current_text = text

        # 1. Credit Card (check before phone/ssn to avoid partial collisions)
        if has_digits and "credit_card" in self.config.pii_types:
            for match in self.PII_PATTERNS["credit_card"].finditer(current_text):
                candidate = match.group(0)
                first_digit = next((c for c in candidate if '0' <= c <= '9'), None)
                if first_digit not in self.VALID_CC_PREFIXES:
                    continue
                if self.luhn_check(candidate):
                    violations.append(
                        ViolationRecord(
                            rule_name="pii_detected_credit_card",
                            risk_level=RiskLevel.HIGH,
                            reason="Valid payment card number (Luhn confirmed) detected in tool response.",
                        )
                    )
                    current_text = current_text.replace(candidate, "[REDACTED_PII:CREDIT_CARD]")

        # 2. SSN
        if has_digits and "ssn" in self.config.pii_types:
            found_ssn = False
            for match in self.PII_PATTERNS["ssn"].finditer(current_text):
                matched_str = match.group(0)
                if "[REDACTED_" not in matched_str:
                    found_ssn = True
                    violations.append(
                        ViolationRecord(
                            rule_name="pii_detected_ssn",
                            risk_level=RiskLevel.HIGH,
                            reason="Social Security Number (SSN) detected in tool response.",
                        )
                    )
            if found_ssn:
                current_text = self.PII_PATTERNS["ssn"].sub("[REDACTED_PII:SSN]", current_text)

        # 3. Email
        if has_at and "email" in self.config.pii_types:
            found_email = False
            for match in self.PII_PATTERNS["email"].finditer(current_text):
                found_email = True
                violations.append(
                    ViolationRecord(
                        rule_name="pii_detected_email",
                        risk_level=RiskLevel.MEDIUM,
                        reason="Email address detected in tool response.",
                    )
                )
            if found_email:
                current_text = self.PII_PATTERNS["email"].sub("[REDACTED_PII:EMAIL]", current_text)

        # 4. Phone
        if has_digits and "phone" in self.config.pii_types:
            found_phone = False
            for match in self.PII_PATTERNS["phone"].finditer(current_text):
                matched_str = match.group(0)
                if not (matched_str.startswith("202") or matched_str.startswith("199")) and "[REDACTED_" not in matched_str:
                    found_phone = True
                    violations.append(
                        ViolationRecord(
                            rule_name="pii_detected_phone",
                            risk_level=RiskLevel.MEDIUM,
                            reason="Telephone number detected in tool response.",
                        )
                    )
            if found_phone:
                current_text = self.PII_PATTERNS["phone"].sub("[REDACTED_PII:PHONE]", current_text)

        return current_text

File: src/core/piiRedactor.js
const { RiskLevel, ViolationRecord } = require('./models');

class PIIRedactor {
  constructor() {
    this.ssnPattern = /\b(\d{3}-\d{2}-\d{4})\b/g;
    this.emailPattern = /\b[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}\b/g;
    this.creditCardPattern = /\b(?:\d[ -]*?){13,19}\b/g;
  }

  luhnCheck(candidate) {
    let clean = '';
    for (let i = 0; i < candidate.length; i++) {
      const code = candidate.charCodeAt(i);
      if (code >= 48 && code <= 57) clean += candidate[i];
    }
    const len = clean.length;
    if (len < 13 || len > 19) return false;
    const first = clean[0];
    if (first !== '2' && first !== '3' && first !== '4' && first !== '5' && first !== '6') return false;

    let sum = 0;
    let shouldDouble = false;
    for (let i = len - 1; i >= 0; i--) {
      let digit = clean.charCodeAt(i) - 48;
      if (shouldDouble) {
        digit *= 2;
        if (digit > 9) digit -= 9;
      }
      sum += digit;
      shouldDouble = !shouldDouble;
    }
    return sum % 10 === 0;
  }

  redact(data) {
    const violations = [];
    const sanitized = this._processData(data, violations);
    return { sanitized, violations };
  }

  _processData(data, violations) {
    if (typeof data === 'string') {
      return this._scanText(data, violations);
    } else if (Array.isArray(data)) {
      return data.map(item => this._processData(item, violations));
    } else if (data && typeof data === 'object') {
      const out = {};
      for (const [k, v] of Object.entries(data)) {
        out[k] = this._processData(v, violations);
      }
      return out;
    }
    return data;
  }

  _scanText(text, violations) {
    let current = text;

    // 1. Credit Cards with Luhn validation
    current = current.replace(this.creditCardPattern, (match) => {
      if (this.luhnCheck(match)) {
        violations.push(new ViolationRecord(
          'pii_detected_credit_card',
          RiskLevel.HIGH,
          'Valid payment card number (Luhn confirmed) detected in tool response.'
        ));
        return '[REDACTED_PII:CREDIT_CARD]';
      }
      return match;
    });

    // 2. SSN
    if (this.ssnPattern.test(current)) {
      violations.push(new ViolationRecord(
        'pii_detected_ssn',
        RiskLevel.HIGH,
        'Social Security Number (SSN) detected in tool response.'
      ));
      current = current.replace(this.ssnPattern, '[REDACTED_PII:SSN]');
    }

    // 3. Email
    if (this.emailPattern.test(current)) {
      violations.push(new ViolationRecord(
        'pii_detected_email',
        RiskLevel.MEDIUM,
        'Email address detected in tool response.'
      ));
      current = current.replace(this.emailPattern, '[REDACTED_PII:EMAIL]');
    }

    return current;
  }
}

module.exports = { PIIRedactor };
